### DISCUSSION TITLE
command line options for xteve

### DISCUSSION THREAD

**Discussion by ReenigneArcher** [streamlink/streamlink#3430] [Q&A] [answered]
Hello,

I discovered streamlink today and so far it is great. It appears to be the best option to play many m3u8 streams that I've had issues playing in vlc and ffmpeg in the past.

So I would like to use streamlink with xteve. Natively xteve can use ffmpeg or vlc, although the software allows you to use a custom path as well. If you're unfamiliar with xteve, it's an m3u proxy. So my end goal is to play the stream in vlc and pass the video back to xteve (without opening any player in the foreground).

So I entered the path of streamlink and the good news is that the command line output from streamlink is passed to xteve. My question is how do I set the command line options?

The default options for xteve are below:
vlc: `vlc.exe -I dummy [URL] --sout #std{mux=ts,access=file,dst=-}`
ffmpeg: `ffmpeg.exe -hide_banner -loglevel error -i [URL] -c copy -f mpegts pipe:1`

I tried this, but it doesn't work. `streamlink.exe --player=vlc --player-args "-I dummy {filename} --sout #std{mux=ts,access=file,dst=-}" "[URL]" best`

I receive the following in the xteve logs:
`2020/12/21 22:26:56 [xTeVe] VLC log:               streamlink: error: unrecognized arguments: --sout# std{mux=ts,access=file,dst=-}" "[URL]" best`

Any idea what the command line should be for this use case?

Edit: To me it doesn't matter if vlc or ffmpeg is the player.

**Comment by back-to** [comment, marked as answer]
It is posible, but you would need to replace FFmpeg or VLC with Streamlink as you already did

> 2020/12/21 22:26:56 [xTeVe] VLC log: streamlink: error: unrecognized arguments: --sout# std{mux=ts,access=file,dst=-}" "[URL]" best

I tested it with the `FFmpeg Binary Path`, but it should also work with `VLC / CVLC Binary Path:`

change `VLC / CVLC Options:` to `--stdout [URL] --default-stream best`

https://streamlink.github.io/cli.html#cmdoption-stdout
https://streamlink.github.io/cli.html#cmdoption-default-stream

If you need more Streamlink settings, you probably need to use Streamlink config files.

https://streamlink.github.io/cli.html#configuration-file
https://streamlink.github.io/cli.html#plugin-specific-configuration-file

---

### Example

working example for `FFmpeg`

Import a Playlist like this and activate it

```m3u
#EXTM3U
#EXTINF:-1,Arte FR
https://www.arte.tv/fr/direct/
#EXTINF:-1,France24
https://www.youtube.com/user/france24
#EXTINF:-1 tvg-id="EuroNews" tvg-name="EuroNews",Euronews
https://www.euronews.com/live
#EXTINF:-1,NTV
https://www.ntv.ru/air/ntv/
```

---

Settings --> Streaming

User Agent: _delete it_

Stream Buffer: `FFmpeg`
FFmpeg Binary Path: `/usr/local/bin/streamlink` will be different on Windows
FFmpeg Options: `--stdout [URL] --default-stream best`

---

after this the URL from the m3u file should work with VLC

http://127.0.0.1:34400/stream/ABC123

**Reply by ReenigneArcher** [reply]
@back-to Okay so for the output options you entered... It appears streamlink isn't trying to use VLC at all.

The command does work, but there is a lot of stuttering. The same stream doesn't stutter at all if I play it from CLI with just `streamlink.exe <myurl> best`

Is it possible to still use VLC but have it in a dummy window and send the output back to streamlink then back to xteve?

**Reply by back-to** [reply]
> It appears streamlink isn't trying to use VLC at all.

Not sure why you would need it,
you get the `http://127.0.0.1:34400/stream/ABC123` url with the Stream which can be used in VLC or some other player.

> The command does work, but there is a lot of stuttering. The same stream doesn't stutter at all if I play it from CLI with just 

share the URL and how you are opening xteve URL.

I open the URL with VLC / MPV and it seems to work normal

> Is it possible to still use VLC but have it in a dummy window and send the output back to streamlink then back to xteve?

you need to reword this, not sure what you want to do here.

**Reply by ReenigneArcher** [reply]
> 
> 
> > It appears streamlink isn't trying to use VLC at all.
> 
> Not sure why you would need it,
> you get the `http://127.0.0.1:34400/stream/ABC123` url with the Stream which can be used in VLC or some other player.
> 
> > The command does work, but there is a lot of stuttering. The same stream doesn't stutter at all if I play it from CLI with just
> 
> share the URL and how you are opening xteve URL.
> 
> I open the URL with VLC / MPV and it seems to work normal
> 
> > Is it possible to still use VLC but have it in a dummy window and send the output back to streamlink then back to xteve?
> 
> you need to reword this, not sure what you want to do here.
I'm going to do more testing to try and figure out what the issue is...

Here is a url I was trying: `https://service-stitcher.clusters.pluto.tv/stitch/hls/channel/591105034c1806b47438342c/master.m3u8?terminate=false&deviceType=web&deviceMake=Chrome&deviceModel=web&sid=115&deviceId=5941303a-0c17-4b7c-bdd6-15b97833afc5&deviceVersion=unknown&appVersion=unknown&clientTime=0&deviceDNT=0&userId=&advertisingId=&appName=web&buildVersion=&appStoreUrl=&architecture=&includeExtendedEvents=false&marketingRegion=US&serverSideAds=true`

xteve is of course changing the url similar to what you entered.

I eventually want to play everything with Plex, but for now just opening the xteve m3u through VLC to reduce complexity at first.

Let's take xteve out of the equation. Say I run from command line `streamlink.exe --stdout <myurl> best`... This is the output I want from VLC, but only after initiated from streamlink. Hope that makes sense.

I am just wanted to test this as xteve reading the --stdout seemed to cause some stuttering, audio sync issues, etc. With xteve natively supporting vlc I was thinking if xteve received the output as it comes vlc it may be better.

If I use the command `streamlink.exe --stdout <myurl> --default-stream best` what is the container of the video? I need everything in .ts

**Reply by back-to** [reply]
> Here is a url I was trying:  https://service-stitcher.clusters.pluto.tv/stitch/hls/channel/591105034c1806b47438342c/master.m3u8? ...

**this won't work**, the stream will break with any ad break.

Update Streamlink and use the Streamlink Plugin `streamlink https://pluto.tv/live-tv/the-asylum best`
Streamlink will use ffmpeg for this Stream and the output should be `mpegts`

> xteve is of course changing the url similar to what you entered.

well, that the proxy between Streamlink and any Videoplayer such as VLC
you can use this URL there.

> I am just wanted to test this as xteve reading the --stdout seemed to cause some stuttering, audio sync issues, etc.

try `Buffer Size:` with **0.5 MB** , I don't have any Audio issues here.
or try some other Streamlink qualitys `[cli][info] Available streams: 572k (worst), 997k, 1539k, 2142k, 3063k (best)`

> With xteve natively supporting vlc, I was thinking if xteve received the output as it comes vlc it may be better.

you can't use VLC as `Stream Buffer:`, if you want to use Streamlink
because Streamlink will be the Player which plays the Stream.

> If I use the command streamlink.exe --stdout <myurl> --default-stream best what is the container of the video? I need everything in .ts

depends, but for your example it should be in _.ts_

---

**xteve** seems pretty slow in opening some Streams

**Reply by ReenigneArcher** [reply]
This issue is actually how I found streamlink (https://github.com/streamlink/streamlink/issues/3156)... everyone seems to have problems with these streams breaking at a commercial break.

The few streams I tested actually got through the commercials fine using the "hls" pluggin, although I didn't test very many cycles. Going to fix my m3u links so it will use the pluto plugin. Thanks!

**Comment by WolfganP** [comment]
I don't know if xteve accepts direct http streaming, but if that's the case using `--player-external-http` may help get rid of vlc/ffmpeg piping/config if there aren't doing any conversion (codec, container) at all.

**Reply by ReenigneArcher** [reply]
For my use case I need all videos in mpegts. Xteve normally accomplishes this by sending the stream urls to ffmpeg or vlc and then reading the output from whichever application is specified. No transcoding is done, but copying to the new container is.
